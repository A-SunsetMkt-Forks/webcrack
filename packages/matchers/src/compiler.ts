import type { NodePath, Visitor } from '@babel/traverse';
import * as t from '@babel/types';
import {
  type ArraySchema,
  type CaptureMatcher,
  type Matcher,
  type NodeSchema,
  type NullableSchema,
  type OrMatcher,
  type Schema,
} from './types.js';

export type CompiledMatcher<T extends t.Node> = (
  input: T,
) => object | undefined;

interface Context {
  captures: string[];
}

/**
 * **Warning**: only compile a schema from trustworthy sources.
 * Otherwise, it could lead to code injection attacks.
 */
export function compile<T extends t.Node>(
  schema: NodeSchema<T>,
): CompiledMatcher<T> {
  const context: Context = { captures: [] };
  const checks = compileNode(schema, 'input', context, false);
  const code = `
const captures = { ${context.captures.map((v) => `${v}: undefined`).join(', ')} };
if (${checks ?? true}) { return captures; }`;

  // eslint-disable-next-line @typescript-eslint/no-implied-eval
  return Function('input', code) as CompiledMatcher<T>;
}

/**
 * **Warning**: only compile a schema from trustworthy sources.
 * Otherwise, it could lead to code injection attacks.
 */
export function compileVisitor<T extends t.Node>(
  schema: NodeSchema<T>,
  cb: (path: NodePath<T>) => void,
  phase: 'enter' | 'exit' = 'enter',
): Visitor {
  const context: Context = { captures: [] };
  const checks = compileNode(schema, 'input', context, false);
  const code = `
const captures = { ${context.captures.map((v) => `${v}: undefined`).join(', ')} };
if (${checks ?? true}) { cb(captures); }`;

  // eslint-disable-next-line @typescript-eslint/no-implied-eval
  const match = Function('input', code) as (
    input: T,
    cb: (path: NodePath<T>) => void,
  ) => void;

  return {
    [schema.type]: {
      [phase](path: NodePath<T>) {
        match(path.node, cb);
      },
    },
  };
}

type CompileFunction = (
  schema: never,
  input: string,
  ctx: Context,
) => string | null;

const compilers: Record<string, CompileFunction> = {
  nullable: compileNullable,
  tuple: compileTuple,
  arrayOf: compileArrayOf,
  or: compileOr,
  capture: compileCapture,
  fromCapture: () => null,
};
Object.keys(t.BUILDER_KEYS).forEach((type) => {
  compilers[type as t.Node['type']] = compileNode;
});

function compileSchema(
  schema: Schema<unknown>,
  input: string,
  ctx: Context,
): string | null {
  if (schema === undefined) {
    return null;
  } else if (
    typeof schema === 'string' ||
    typeof schema === 'number' ||
    typeof schema === 'boolean' ||
    schema === null
  ) {
    return `${input} === ${JSON.stringify(schema)}`;
  } else if (Array.isArray(schema)) {
    return compileTuple(schema, input, ctx);
  } else {
    return compileMatcher(schema as Matcher<unknown>, input, ctx);
  }
}

function compileMatcher(
  matcher: Matcher<unknown>,
  input: string,
  ctx: Context,
): string | null {
  return compilers[matcher.type](matcher as never, input, ctx);
}

function compileTuple(
  schema: Schema<unknown>[],
  input: string,
  ctx: Context,
): string | null {
  const checks = [
    `${input}.length === ${schema.length}`,
    ...schema
      .map((m, i) => compileSchema(m, `${input}[${i}]`, ctx))
      .filter((c) => c !== null),
  ];
  return checks.join(' && ');
}

function compileArrayOf(
  schema: ArraySchema<unknown>,
  input: string,
  ctx: Context,
): string | null {
  const checks = compileSchema(schema.schema, 'element', ctx);
  if (checks === null) return null;
  return `${input}.every(element => ${checks})`;
}

function compileNullable(
  schema: NullableSchema<unknown>,
  input: string,
  ctx: Context,
): string | null {
  const checks = compileSchema(schema.schema, input, ctx);
  if (checks === null) return null;
  return `(${input} === null || ${checks})`;
}

function compileOr(schema: OrMatcher<unknown[]>, input: string, ctx: Context) {
  const checks = schema.schema
    .map((m) => compileSchema(m, input, ctx))
    .filter((c) => c !== null);
  if (checks.length === 0) return null;
  return `(${checks.join(' || ')})`;
}

function compileCapture(
  schema: CaptureMatcher<unknown>,
  input: string,
  ctx: Context,
) {
  ctx.captures.push(schema.name);
  const checks = compileSchema(schema.schema, input, ctx);
  if (checks === null) {
    return `(captures['${schema.name}'] = ${input}, true)`;
  } else {
    return `${checks} && (captures['${schema.name}'] = ${input}, true)`;
  }
}

function compileNode(
  schema: NodeSchema<t.Node>,
  input: string,
  ctx: Context,
  checkType = true,
) {
  const checks: string[] = [];
  if (checkType) {
    checks.push(`${input}.type === '${schema.type}'`);
  }

  const fields = t.NODE_FIELDS[schema.type];

  for (const key in fields) {
    const value = schema[key as never] as Schema<unknown>;
    const valueChecks = compileSchema(value, `${input}.${key}`, ctx);
    if (valueChecks === null) continue;

    const validator = fields[key];
    if (value !== null && validator.optional) {
      // TODO: only when all checks includes property accesses
      checks.push(`${input}.${key} != null`);
    }
    checks.push(valueChecks);
  }

  return checks.length ? checks.join(' && ') : null;
}
