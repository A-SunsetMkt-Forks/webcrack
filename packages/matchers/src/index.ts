export * from './compiler.js';
export * from './generated.js';
export * from './types.js';
