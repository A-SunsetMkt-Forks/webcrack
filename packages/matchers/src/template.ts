import { parse, parseExpression, type ParserOptions } from '@babel/parser';
import * as t from '@babel/types';
import { type NodeSchema } from './types.js';

export function expression(
  strings: TemplateStringsArray,
  ...schemas: NodeSchema<t.Node>[]
): NodeSchema<t.Expression> {
  return parseTemplate(
    strings,
    schemas,
    parseExpression,
  ) as NodeSchema<t.Expression>;
}

export function statement(
  strings: TemplateStringsArray,
  ...schemas: NodeSchema<t.Statement>[]
): NodeSchema<t.Statement> {
  return parseTemplate(
    strings,
    schemas,
    parseStatement,
  ) as NodeSchema<t.Statement>;
}

function parseStatement(input: string) {
  return parse(input, { allowReturnOutsideFunction: true }).program.body[0];
}

function parseTemplate(
  strings: TemplateStringsArray,
  schemas: NodeSchema<t.Node>[],
  parse: (input: string, options?: ParserOptions) => t.Node,
): NodeSchema<t.Node> {
  let schemaIndex = 0;
  const pattern = strings.reduce((acc, curr, i) => {
    acc += curr;

    if (schemas[i]) {
      acc += `$${schemaIndex++}`;
    }

    return acc;
  }, '');

  const ast = parse(pattern);
  let rootSchema: NodeSchema<t.Node> | undefined;
  t.traverse(ast, {
    enter(node, ancestors) {
      if (!t.isIdentifier(node)) return;
      if (!/^\$\d+$/.test(node.name)) return;
      const matcher = schemas[Number(node.name.slice(1))];
      const ancestor = ancestors.at(-1);
      if (ancestor) {
        const container = ancestor.node[ancestor.key as keyof t.Node];
        if (Array.isArray(container)) {
          container[ancestor.index!] = matcher as never;
        } else {
          ancestor.node[ancestor.key as keyof t.Node] = matcher as never;
        }
      } else {
        rootSchema = matcher;
      }
    },
  });

  return rootSchema ?? ast;
}
