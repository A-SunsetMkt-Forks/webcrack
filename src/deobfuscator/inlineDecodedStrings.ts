import * as t from '@babel/types';
import { AsyncTransform } from '../transforms';
import { VMDecoder } from './vm';

/**
 * Replaces calls to decoder functions with the decoded string.
 * E.g. `m(199)` -> `'log'`
 */
export default {
  name: 'inlineDecodedStrings',
  tags: ['unsafe'],
  async run(ast, state, options) {
    if (!options) return;

    const calls = options.vm.decoders.flatMap(decoder =>
      decoder.collectCalls()
    );
    const decodedValues = await options.vm.decode(calls);

    for (let i = 0; i < calls.length; i++) {
      const call = calls[i];
      const value = decodedValues[i];

      call.replaceWith(t.valueToNode(value));
      if (typeof value !== 'string')
        call.addComment('leading', 'webcrack:decode_error');
    }

    state.changes += calls.length;
  },
} satisfies AsyncTransform<{ vm: VMDecoder }>;
