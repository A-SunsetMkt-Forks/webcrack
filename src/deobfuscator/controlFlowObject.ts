import { Binding, NodePath } from '@babel/traverse';
import * as t from '@babel/types';
import { FunctionExpression } from '@babel/types';
import * as m from '@codemod/matchers';
import { Transform } from '../transforms';
import { getPropName } from '../utils/ast';
import { inlineCfFunction } from '../utils/inline';
import {
  constKey,
  constMemberExpression,
  createFunctionMatcher,
  findParent,
  isReadonlyObject,
} from '../utils/matcher';
import { renameFast } from '../utils/rename';

/**
 * Explanation: https://excalidraw.com/#json=0vehUdrfSS635CNPEQBXl,hDOd-UO9ETfSDWT9MxVX-A
 */

export default {
  name: 'controlFlowObject',
  tags: ['safe'],
  visitor() {
    const varId = m.capture(m.identifier());
    const propertyName = m.matcher<string>(i =>
      /^[a-z]{5}$/i.test(i as string)
    );
    const propertyKey = constKey(propertyName);
    const property = m.or(
      // E.g. "6|0|4|3|1|5|2"
      m.stringLiteral(),
      // E.g. function (a, b) { return a + b }
      createFunctionMatcher(2, (left, right) => [
        m.returnStatement(
          m.or(
            m.binaryExpression(undefined, left, right),
            m.logicalExpression(undefined, left, right)
          )
        ),
      ]),
      // E.g. function (a, b, c) { return a(b, c) } with an arbitrary number of arguments
      m.matcher<FunctionExpression>(node => {
        return (
          t.isNode(node) &&
          t.isFunctionExpression(node) &&
          createFunctionMatcher(node.params.length, (...params) => [
            m.returnStatement(m.callExpression(params[0], params.slice(1))),
          ]).match(node)
        );
      })
    );
    // E.g. "rLxJs": "6|0|4|3|1|5|2"
    const objectProperties = m.capture(
      m.arrayOf(m.objectProperty(propertyKey, property))
    );
    const aliasId = m.capture(m.identifier());
    const aliasVar = m.variableDeclarator(aliasId, m.fromCapture(varId));
    // E.g. "rLxJs"
    const assignedKey = m.capture(propertyName);
    // E.g. "6|0|4|3|1|5|2"
    const assignedValue = m.capture(property);
    // E.g. obj.rLxJs = "6|0|4|3|1|5|2"
    const assignment = m.expressionStatement(
      m.assignmentExpression(
        '=',
        constMemberExpression(m.fromCapture(varId), assignedKey),
        assignedValue
      )
    );
    // E.g. obj.rLxJs
    const memberAccess = constMemberExpression(
      m.or(m.fromCapture(varId), m.fromCapture(aliasId)),
      propertyName
    );
    const varMatcher = m.variableDeclarator(
      varId,
      m.capture(m.objectExpression(objectProperties))
    );

    function isConstantBinding(binding: Binding) {
      // Workaround because sometimes babel treats the VariableDeclarator/binding itself as a violation
      return binding.constant || binding.constantViolations[0] === binding.path;
    }

    function transform(path: NodePath<t.VariableDeclarator>) {
      let changes = 0;
      if (varMatcher.match(path.node)) {
        // Verify all references to make sure they match how the obfuscator
        // would have generated the code (no reassignments, etc.)
        const binding = path.scope.getBinding(varId.current!.name);
        if (!binding) return changes;
        if (!isConstantBinding(binding)) return changes;
        if (objectProperties.current!.length === 0)
          transformObjectKeys(binding);
        if (!isReadonlyObject(binding, memberAccess)) return changes;

        const props = new Map(
          objectProperties.current!.map(p => [
            getPropName(p.key),
            p.value as t.FunctionExpression | t.StringLiteral,
          ])
        );
        if (!props.size) return changes;

        const oldRefs = [...binding.referencePaths];

        // Have to loop backwards because we might replace a node that
        // contains another reference to the binding (https://github.com/babel/babel/issues/12943)
        [...binding.referencePaths].reverse().forEach(ref => {
          const memberPath = ref.parentPath as NodePath<t.MemberExpression>;
          const propName = getPropName(memberPath.node.property)!;
          const value = props.get(propName)!;

          if (t.isStringLiteral(value)) {
            memberPath.replaceWith(value);
          } else {
            inlineCfFunction(
              value,
              memberPath.parentPath as NodePath<t.CallExpression>
            );
          }
          changes++;
        });

        oldRefs.forEach(ref => {
          const varDeclarator = findParent(ref, m.variableDeclarator());
          if (varDeclarator) changes += transform(varDeclarator);
        });

        path.remove();
        changes++;
      }
      return changes;
    }

    /**
     * When the `Transform Object Keys` option is enabled, the obfuscator generates an empty
     * object, assigns the properties later and adds an alias variable to the object.
     * This function undoes that by converting the assignments to inline object properties.
     */
    function transformObjectKeys(objBinding: Binding) {
      const refs = objBinding.referencePaths;

      if (refs.length < 2) return;
      if (!aliasVar.match(refs.at(-1)?.parent)) return;

      const assignments: NodePath[] = [];

      for (let i = 0; i < refs.length - 1; i++) {
        const expressionStatement = refs[i].parentPath?.parentPath?.parentPath;
        if (!assignment.match(expressionStatement?.node)) return;

        assignments.push(expressionStatement!);
        objectProperties.current!.push(
          t.objectProperty(
            t.identifier(assignedKey.current!),
            assignedValue.current!
          )
        );
      }

      const aliasBinding = objBinding.scope.getBinding(aliasId.current!.name)!;
      if (!isReadonlyObject(aliasBinding, memberAccess)) return;

      objBinding.referencePaths = aliasBinding.referencePaths;
      objBinding.references = aliasBinding.references;

      renameFast(aliasBinding, objBinding.identifier.name);

      assignments.forEach(p => p.remove());
      aliasBinding.path.remove();
    }

    return {
      VariableDeclarator: {
        exit(path) {
          this.changes += transform(path);
        },
      },
      noScope: true,
    };
  },
} satisfies Transform;
