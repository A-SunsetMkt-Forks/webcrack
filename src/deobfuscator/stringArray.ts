import traverse, { NodePath } from '@babel/traverse';
import * as t from '@babel/types';
import * as m from '@codemod/matchers';
import { renameFast } from '../utils/rename';

export interface StringArray {
  path: NodePath<t.FunctionDeclaration>;
  references: NodePath[];
  name: string;
  length: number;
}

export function findStringArray(ast: t.Node): StringArray | undefined {
  let result: StringArray | undefined;
  const functionName = m.capture(m.anyString());
  const arrayIdentifier = m.capture(m.identifier());
  const arrayExpression = m.capture(
    m.arrayExpression(m.arrayOf(m.stringLiteral()))
  );
  // getStringArray = function () { return n; };
  const functionAssignment = m.assignmentExpression(
    '=',
    m.identifier(m.fromCapture(functionName)),
    m.functionExpression(
      undefined,
      [],
      m.blockStatement([m.returnStatement(m.fromCapture(arrayIdentifier))])
    )
  );
  const variableDeclaration = m.variableDeclaration(undefined, [
    m.variableDeclarator(arrayIdentifier, arrayExpression),
  ]);
  // function getStringArray() { ... }
  const matcher = m.functionDeclaration(
    m.identifier(functionName),
    [],
    m.or(
      // var array = ["hello", "world"];
      // return (getStringArray = function () { return array; })();
      m.blockStatement([
        variableDeclaration,
        m.returnStatement(m.callExpression(functionAssignment)),
      ]),
      // var array = ["hello", "world"];
      // getStringArray = function () { return n; });
      // return getStringArray();
      m.blockStatement([
        variableDeclaration,
        m.expressionStatement(functionAssignment),
        m.returnStatement(m.callExpression(m.identifier(functionName))),
      ])
    )
  );

  traverse(ast, {
    FunctionDeclaration(path) {
      if (matcher.match(path.node)) {
        const length = arrayExpression.current!.elements.length;
        const name = functionName.current!;
        const binding = path.scope.getBinding(name)!;
        renameFast(binding, '__STRING_ARRAY__');

        result = {
          path,
          references: binding.referencePaths,
          name: '__STRING_ARRAY__',
          length,
        };
        path.stop();
      }
    },
  });

  return result;
}
