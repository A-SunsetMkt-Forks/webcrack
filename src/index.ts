import generate from '@babel/generator';
import { parse } from '@babel/parser';
import * as m from '@codemod/matchers';
import debug from 'debug';
import { join } from 'node:path';
import deobfuscator from './deobfuscator';
import debugProtection from './deobfuscator/debugProtection';
import selfDefending from './deobfuscator/selfDefending';
import {
  Sandbox,
  createBrowserSandbox,
  createNodeSandbox,
} from './deobfuscator/vm';
import { unpackBundle } from './extractor';
import { Bundle } from './extractor/bundle';
import {
  applyTransform,
  applyTransformAsync,
  applyTransforms,
} from './transforms';
import blockStatement from './transforms/blockStatement';
import jsx from './transforms/jsx';
import sequence from './transforms/sequence';
import splitVariableDeclarations from './transforms/splitVariableDeclarations';
import unminify from './transforms/unminify';

export interface WebcrackResult {
  code: string;
  bundle: Bundle | undefined;
  /**
   * Save the deobufscated code and the extracted bundle to the given directory.
   * @param path Output directory
   */
  save(path: string): Promise<void>;
}

export interface Options {
  /**
   * Decompile react components to JSX.
   * @default true
   */
  jsx?: boolean;
  /**
   * Extract modules from the bundle.
   * @default true
   */
  unpack?: boolean;
  /**
   * Deobfuscate the code.
   * @default true
   */
  deobfuscate?: boolean;
  /**
   * Assigns paths to modules based on the given matchers.
   * This will also rewrite `require()` calls to use the new paths.
   *
   * @example
   * ```js
   * m => ({
   *   './utils/color.js': m.regExpLiteral('^#([0-9a-f]{3}){1,2}$')
   * })
   * ```
   */
  mappings?: (
    m: typeof import('@codemod/matchers')
  ) => Record<string, m.Matcher<unknown>>;
  /**
   * Function that executes a code expression and returns the result (typically from the obfuscator).
   */
  sandbox?: Sandbox;
}

function mergeOptions(options: Options): asserts options is Required<Options> {
  const mergedOptions: Required<Options> = {
    jsx: true,
    unpack: true,
    deobfuscate: true,
    mappings: () => ({}),
    sandbox: process.env.browser ? createBrowserSandbox() : createNodeSandbox(),
    ...options,
  };
  Object.assign(options, mergedOptions);
}

export async function webcrack(
  code: string,
  options: Options = {}
): Promise<WebcrackResult> {
  mergeOptions(options);

  if (process.env.browser) {
    debug.enable('webcrack:*');
  }

  const ast = parse(code, {
    sourceType: 'unambiguous',
    allowReturnOutsideFunction: true,
    plugins: ['jsx'],
  });

  applyTransforms(
    ast,
    [blockStatement, sequence, splitVariableDeclarations],
    'prepare'
  );

  if (options.deobfuscate)
    await applyTransformAsync(ast, deobfuscator, options.sandbox);

  applyTransform(ast, unminify);

  // TODO: Also merge unminify visitor (breaks selfDefending/debugProtection atm)
  applyTransforms(
    ast,
    [
      // Have to run this after unminify to properly detect it
      options.deobfuscate ? [selfDefending, debugProtection] : [],
      options.jsx ? [jsx] : [],
    ].flat()
  );

  // Unpacking modifies the same AST and may result in imports not at top level
  // so the code has to be generated before
  const outputCode = generate(ast).code;

  const bundle = options.unpack
    ? unpackBundle(ast, options.mappings(m))
    : undefined;

  return {
    code: outputCode,
    bundle,
    async save(path) {
      if (process.env.browser) {
        throw new Error('Not implemented.');
      } else {
        const { mkdir, writeFile } = await import('node:fs/promises');
        await mkdir(path, { recursive: true });
        await writeFile(join(path, 'deobfuscated.js'), outputCode, 'utf8');
        await bundle?.save(path);
      }
    },
  };
}
