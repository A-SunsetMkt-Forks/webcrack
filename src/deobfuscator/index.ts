import debug from 'debug';
import {
  applyTransform,
  applyTransformAsync,
  applyTransforms,
  AsyncTransform,
} from '../transforms';
import mergeStrings from '../transforms/mergeStrings';
import { codePreview } from '../utils/ast';
import { findArrayRotator } from './arrayRotator';
import controlFlowObject from './controlFlowObject';
import controlFlowSwitch from './controlFlowSwitch';
import deadCode from './deadCode';
import { findDecoders } from './decoder';
import inlineDecodedStrings from './inlineDecodedStrings';
import inlineDecoderWrappers from './inlineDecoderWrappers';
import objectLiterals from './objectLiterals';
import { findStringArray } from './stringArray';
import { Sandbox, VMDecoder } from './vm';

// https://astexplorer.net/#/gist/b1018df4a8daebfcb1daf9d61fe17557/4ff9ad0e9c40b9616956f17f59a2d9888cd62a4f

export default {
  name: 'deobfuscate',
  tags: ['unsafe'],
  async run(ast, state, sandbox) {
    const logger = debug('webcrack:deobfuscate');
    if (!sandbox) return;

    const stringArray = findStringArray(ast);
    logger(`String Array: ${stringArray ? 'yes' : 'no'}`);
    if (!stringArray) return;
    logger(` - length: ${stringArray.length}`);
    logger(` - ${codePreview(stringArray.path.node)}`);

    const rotator = findArrayRotator(stringArray);
    logger(`String Array Rotate: ${rotator ? 'yes' : 'no'}`);
    if (rotator) {
      logger(` - ${codePreview(rotator.node)}`);
    }

    const decoders = findDecoders(stringArray);
    logger(`String Array Encodings: ${decoders.length}`);

    state.changes += applyTransform(ast, objectLiterals).changes;

    for (const decoder of decoders) {
      logger(` - ${codePreview(decoder.path.node)}`);

      state.changes += applyTransform(
        ast,
        inlineDecoderWrappers,
        decoder.path
      ).changes;
    }

    const vm = new VMDecoder(sandbox, stringArray, decoders, rotator);
    state.changes += (
      await applyTransformAsync(ast, inlineDecodedStrings, { vm })
    ).changes;

    stringArray.path.remove();
    rotator?.remove();
    decoders.forEach(decoder => decoder.path.remove());
    state.changes += 2 + decoders.length;

    state.changes += applyTransforms(
      ast,
      [mergeStrings, deadCode, controlFlowObject, controlFlowSwitch],
      'mergeStrings, deadCode, controlFlow'
    ).changes;
  },
} satisfies AsyncTransform<Sandbox>;
