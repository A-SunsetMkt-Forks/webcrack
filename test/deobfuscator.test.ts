import generate from '@babel/generator';
import { parse } from '@babel/parser';
import traverse from '@babel/traverse';
import assert from 'assert';
import { readFile } from 'fs/promises';
import { join } from 'node:path';
import { describe, expect, test } from 'vitest';
import { findStringArray } from '../src/deobfuscator/stringArray';
import { webcrack } from '../src/index';
import {
  inlineFunctionAliases,
  inlineVariableAliases,
} from '../src/utils/inline';

// Test samples
test.each([
  'obfuscator.io.js',
  'obfuscator.io-rotator-unary.js',
  'obfuscator.io-multi-encoders.js',
  'obfuscator.io-function-wrapper.js',
  'obfuscator.io-calls-transform.js',
  'obfuscator.io-control-flow.js',
  'obfuscator.io-control-flow-keys.js',
  'obfuscator.io-high.js',
])('deobfuscate %s', async filename => {
  const result = await webcrack(
    await readFile(join('test', 'samples', filename), 'utf8')
  );
  expect(result.code).toMatchSnapshot();
});

describe('find string array', async () => {
  const ast = parse(await readFile('./test/samples/obfuscator.io.js', 'utf8'));

  test('function wrapper', () => {
    const stringArray = findStringArray(ast);
    expect(stringArray).toBeDefined();
    assert(stringArray);
    expect(stringArray.name).toBe('__STRING_ARRAY__');
    expect(stringArray.references).toHaveLength(3);
    expect(stringArray.length).toBe(25);
  });
});

describe('inline decoder', () => {
  test('inline variable', () => {
    const ast = parse(`
      function decoder() {}
      decoder(1);
      (() => {
        const alias = decoder, alias3 = alias;
        alias(2);
        alias3(3);
        (() => {
          let alias2;
          (alias2 = alias)(4);
        });
        let alias4;
        alias4 = alias;
        alias4(5);
      });
  `);
    traverse(ast, {
      FunctionDeclaration(path) {
        const binding = path.scope.getBinding('decoder')!;
        inlineVariableAliases(binding);
        path.stop();
      },
    });
    expect(generate(ast).code).toMatchSnapshot();
  });

  test('inline function', () => {
    const ast = parse(`
      function decoder() {}
      decoder(1);
      (() => {
        function alias(a, b) {
          return decoder(a - 625, b);
        }
        alias(2, 3);
        (() => {
          function alias2(a, b) {
            return alias(b - -678, a);
          }
          alias2(4, 5);
        })();
      })();
  `);
    traverse(ast, {
      FunctionDeclaration(path) {
        const binding = path.scope.parent.bindings.decoder;
        inlineFunctionAliases(binding);
        path.stop();
      },
    });

    expect(generate(ast).code).toMatchSnapshot();
  });
});
