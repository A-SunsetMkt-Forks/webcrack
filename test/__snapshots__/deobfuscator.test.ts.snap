// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deobfuscate obfuscator.io.js 1`] = `
"// v4.0.0

function hi() {
  console.log(\\"Hello World!\\");
  console.log( /*webcrack:decode_error*/undefined);
  console.log( /*webcrack:decode_error*/undefined);
  const p = true;
  console.log(p ? \\"log\\" : \\"Hello World!\\");
}
hi();"
`;

exports[`deobfuscate obfuscator.io-calls-transform.js 1`] = `
"function foo() {
  var c = \\"string1\\";
  var d = \\"string2\\";
  var e = \\"string3\\";
  var f = \\"string4\\";
  var g = \\"string5\\";
  var h = \\"string6\\";
}
console.log(foo());"
`;

exports[`deobfuscate obfuscator.io-control-flow.js 1`] = `
"function applyTransforms() {
  this.modules.forEach(varInjection_1.inlineVarInjections);
  this.modules.forEach(esm_1.convertESM);
  (0, getDefaultExport_1.convertDefaultRequire)(this);
  this.replaceRequireCalls();
  var f = m.capture(m.numericLiteral());
  var g = m.callExpression(m.identifier(\\"require\\"), [f]);
  return g;
}"
`;

exports[`deobfuscate obfuscator.io-control-flow-keys.js 1`] = `
"function hi() {
  const _0x46643e = 1;
  const _0x4d295f = 2;
  if (_0x46643e < _0x4d295f) {
    console.log(\\"Hello World!\\");
  }
}
hi();"
`;

exports[`deobfuscate obfuscator.io-function-wrapper.js 1`] = `
"function hi() {
  console.log(\\"Hello World!\\");
  console.log(30);
  console.log(undefined);
}
hi();"
`;

exports[`deobfuscate obfuscator.io-high.js 1`] = `
"function hi() {
  console.log(\\"Hello World!\\");
}
hi();"
`;

exports[`deobfuscate obfuscator.io-multi-encoders.js 1`] = `
"function hi() {
  console.log(\\"Hello World!\\");
}
hi();"
`;

exports[`deobfuscate obfuscator.io-rotator-unary.js 1`] = `
"function hi() {
  console.log(\\"Hello World!\\");
}
hi();"
`;

exports[`inline decoder > inline function 1`] = `
"function decoder() {}
decoder(1);
(() => {
  decoder(2 - 625, 3);
  (() => {
    decoder(5 - -678 - 625, 4);
  })();
})();"
`;

exports[`inline decoder > inline variable 1`] = `
"function decoder() {}
decoder(1);
() => {
  decoder(2);
  decoder(3);
  () => {
    alias(4);
  };
  decoder(5);
};"
`;
